code/modules/day_night/day_night_controller.dm

```diff
@@ -0,0 +1,253 @@
+ * Returns TRUE if it was successfully registered, otherwise FALSE if not.
+ */
+/datum/day_night_controller/proc/register_unaffected_area(area/area_to_register)
+	if(area_to_register in unaffected_area_cache)
+		return FALSE
+	unaffected_area_cache += area_to_register
+	area_to_register.update_day_night_turfs(TRUE, FALSE, src)
+	area_to_register.RegisterSignal(src, COMSIG_DAY_NIGHT_CONTROLLER_LIGHT_UPDATE, TYPE_PROC_REF(/area, apply_day_night_turfs))
+	RegisterSignal(area_to_register, COMSIG_PARENT_QDELETING, PROC_REF(unregister_unaffected_area))
+	RegisterSignal(area_to_register, COMSIG_AREA_AFTER_SHUTTLE_MOVE, PROC_REF(after_shuttle_move))
+	return TRUE
+
+/**
+ * Removes an unaffected turf from the controller and clears up references.
+ * * Arguments:
+ * * area_to_unregister - The area that is to be unregistered from this controller
+ */
+/datum/day_night_controller/proc/unregister_unaffected_area(area/area_to_unregister)
+	SIGNAL_HANDLER
+
+	area_to_unregister.clear_adjacent_turfs()
+	area_to_unregister.UnregisterSignal(src, COMSIG_DAY_NIGHT_CONTROLLER_LIGHT_UPDATE)
+	UnregisterSignal(area_to_unregister, COMSIG_AREA_AFTER_SHUTTLE_MOVE)
+	UnregisterSignal(area_to_unregister, COMSIG_PARENT_QDELETING)
+
+	unaffected_area_cache -= area_to_unregister
+
+/**
+ * Called after a shuttle moves from a registered area, checks to see if the new level is different from ours.
+ */
+/datum/day_night_controller/proc/after_shuttle_move(area/area_to_check)
+	SIGNAL_HANDLER
+
+	if(area_to_check.z == affected_z_level)
+		return
+
+	unregister_unaffected_area(area_to_check)
+
+
+/**
+ * Creates all of the desired lightzone datums and caches them for later.
+ */
+/datum/day_night_controller/proc/load_lightzones()
+	for(var/iterating_lightzone_type in lightzones)
+		lightzone_cache += new iterating_lightzone_type
+
+/**
+ * Gets the lightzone that relates to the given hour
+ * Arguments:
+ * * hour - The hour at which we will be checking our lightzones.
+ * Returns the lightzone datum.
+ */
+/datum/day_night_controller/proc/get_lightzone(hour)
+	var/lightzone_to_return
+	if(hour == MIDNIGHT_RESET)
+		hour = 0
+	for(var/datum/lightzone/iterating_lightzone as anything in lightzone_cache)
+		if((hour >= iterating_lightzone.start_hour) && (hour < (!iterating_lightzone.end_hour ? 24 : iterating_lightzone.end_hour)))
+			lightzone_to_return = iterating_lightzone
+	if(!lightzone_to_return)
+		CRASH("Error while finding a lightzone in slot [hour] for [type]!")
+	return lightzone_to_return
+
+/**
+ * Applys the current lightzone to all of the areas according to the lookup tables and hour.
+ * Arguments:
+ * * hour - The index to use in our lookup tables.
+ */
+/datum/day_night_controller/proc/update_time(hour)
+	if(color_lookup_table["[hour ? hour - 1 : 24]"] == color_lookup_table["[hour]"] && alpha_lookup_table["[hour ? hour - 1 : 24]"] == alpha_lookup_table["[hour]"]) // Why change the color of the light when it's the same as the last?????
```

Comment:
We do this 0-24 thing so much that it should be a proc for saturating subtract/add/whatever.